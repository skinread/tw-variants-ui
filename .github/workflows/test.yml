# This workflow will do a clean installation of node dependencies, cache/restore them, build the source code and run tests across different versions of node
# For more information see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-nodejs

name: CI Tests

on:
  pull_request:
    branches: ["main","develop"]

jobs:
  test:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        node-version: [20.x]
    steps:
    - name: Checkout
      uses: actions/checkout@v4
    - name: Setup Node ${{ matrix.node-version }}
      uses: actions/setup-node@v4
      with:
        node-version: ${{ matrix.node-version }}
        cache: 'npm'
        cache-dependency-path: "**/package-lock.json"
    - name: Cache or restore dependencies
      uses: actions/cache@v4
      with:
        path: ~/.npm
        key: ${{ runner.os }}-npm-${{ hashFiles('**/package-lock.json') }}
        restore-keys: ${{ runner.os }}-npm-
    - name: Install dependencies
      run: npm ci

    - name: Check formatting and linting
      run: npm run prettier && npm run lint

    - name: Check package build
      run: npm run build

    - name: Get Playwright version
      run: echo "PLAYWRIGHT_VERSION=$(jq -r '.devDependencies["@playwright/test"]' package.json)" >> $GITHUB_ENV
    - name: Cache or restore Playwright binaries
      uses: actions/cache@v4
      with:
        path: |
          ~/.cache/ms-playwright
        key: ${{ runner.os }}-playwright-${{ env.PLAYWRIGHT_VERSION }}
        restore-keys: ${{ runner.os }}-playwright-
    - name: Install Playwright
      if: steps.playwright-cache.outputs.cache-hit != 'true'
      run: npm run pretest
    - name: Storybook test build
      run: npm run storybook:build -- --test
    - name: Storybook test suite
      run: npm run test:ci
