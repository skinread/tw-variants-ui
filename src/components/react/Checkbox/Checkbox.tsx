import { forwardRef, useRef } from 'react';
import { inputVariants, validityFromProps } from '@common';
import { type CommonFieldProps } from '@components/react';
import { FieldFeedback } from '../InputField/FieldFeedback';

type ComponentProps = Omit<React.ComponentPropsWithoutRef<'input'>, 'name'>;

type CommonFieldPropsFiltered = Pick<CommonFieldProps, 'name' | 'hasError' | 'feedbackContent'>;

export interface CheckboxProps extends CommonFieldPropsFiltered, ComponentProps {
  /**
   * Associate a descriptive label with this form element, required for accessibility. Usually a plain string, but also handles JSX content for links, etc.
   */
  label: React.ReactNode;
  /**
   * Pass the `checked` prop for the input to be controlled by React, this will require an `onClick` event handler
   */
  checked?: ComponentProps['checked'];
}

/**
 * Styled checkbox input field. React `ref` enabled. Also exposes intrinsic `<input>` attributes.
 */
export const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>(
  function VuiCheckbox(props, ref) {
    const { feedbackContent, hasError, id: idProp, label, ...elemAttrs } = props;
    const {
      wrapper: wrapperStyle,
      label: labelStyle,
      labelText: labelTextStyle,
      input: inputStyle,
    } = inputVariants({ type: 'checkbox', validity: validityFromProps({ hasError }) });

    const id = useRef(idProp ?? `checkbox-${props.name}`);
    const idFeedback = `${id.current}-feedback`;

    return (
      <div className={wrapperStyle()}>
        <div className="flex">
          <label className={labelStyle()}>
            <input type="checkbox" className={inputStyle()} {...elemAttrs} ref={ref} />
            <span className={labelTextStyle()}>{label}</span>
          </label>
        </div>
        <FieldFeedback
          feedbackContent={feedbackContent}
          idFeedback={idFeedback}
          hasError={hasError}
        />
      </div>
    );
  }
);
