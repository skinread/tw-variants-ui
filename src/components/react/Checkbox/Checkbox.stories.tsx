import type { Meta, StoryObj } from '@storybook/react';
import { expect, fn, userEvent, within } from '@storybook/test';

import { Checkbox } from './Checkbox';

const meta = {
  title: 'React/Checkbox',
  component: Checkbox,
  tags: ['autodocs'],
  args: {
    label: 'Yes I agree',
    name: 'demo-checkbox',
    checked: undefined,
    hasError: undefined,
    onChange: fn(),
  },
} satisfies Meta<typeof Checkbox>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Basic: Story = {};

export const LabelWithLink: Story = {
  args: {
    label: (
      <>
        Yes I agree to the{' '}
        <a href="#link" className="link">
          terms and conditions
        </a>
      </>
    ),
    name: 'demo-label-link-checkbox',
  },
};

export const Invalid: Story = {
  args: {
    ...LabelWithLink.args,
    name: 'demo-checkbox-invalid',
    feedbackContent: 'You must agree',
    hasError: true,
  },
};

/**
 * Technical reference: `onChange` handler required when the state is controlled and the `checked` prop should be derived from state
 */
export const ControlledChecked: Story = {
  args: {
    label: 'Can\'t click me. Toggle me in React using the "checked" prop',
    name: 'demo-controlled-checkbox',
    checked: true,
  },
};

/** An storybook story with an interaction test for a input checkbox component */
export const InteractionTest: Story = {
  args: {
    ...LabelWithLink.args,
  },
  play: async ({ args, canvasElement, step }) => {
    const canvas = within(canvasElement);
    const input = canvas.getByRole('checkbox');
    const label = input.parentElement;

    await step('Initial state', async () => {
      await expect(label).toHaveTextContent('Yes I agree');
      await expect(input).not.toBeChecked();
      await expect(input).toHaveAttribute('name', args.name);
    });

    await step('Checked state', async () => {
      await userEvent.click(label ?? input, { delay: 100 });
      await expect(args.onChange).toHaveBeenCalled();
      await expect(input).toBeChecked();
    });

    await step('Keyboard interaction', async () => {
      await userEvent.keyboard('[Space]', { delay: 100 });
      await expect(args.onChange).toHaveBeenCalled();
      await expect(input).not.toBeChecked();
    });

    await step('Link in label', async () => {
      const link = canvas.getByRole('link');
      await expect(link).toHaveTextContent('terms and conditions');
      await expect(link).toHaveAttribute('href', '#link');
    });
  },
};
