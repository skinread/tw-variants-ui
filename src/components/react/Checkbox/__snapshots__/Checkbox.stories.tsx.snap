// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React/Checkbox Basic smoke-test 1`] = `
<div class="form-control block my-4">
  <div class="flex">
    <label class="label items-start px-0 py-0 cursor-pointer">
      <input type="checkbox"
             class="placeholder:text-subtle checkbox checkbox-primary w-[--vui-checkbox-size] h-[--vui-checkbox-size] rounded-[--rounded-field] bg-white"
             name="demo-checkbox"
      >
      <span class="label-text text-base ml-2.5">
        Yes I agree
      </span>
    </label>
  </div>
</div>
`;

exports[`React/Checkbox ControlledChecked smoke-test 1`] = `
<div class="form-control block my-4">
  <div class="flex">
    <label class="label items-start px-0 py-0 cursor-pointer">
      <input type="checkbox"
             class="placeholder:text-subtle checkbox checkbox-primary w-[--vui-checkbox-size] h-[--vui-checkbox-size] rounded-[--rounded-field] bg-white"
             name="demo-controlled-checkbox"
             checked
      >
      <span class="label-text text-base ml-2.5">
        Can't click me. Toggle me in React using the "checked" prop
      </span>
    </label>
  </div>
</div>
`;

exports[`React/Checkbox Invalid smoke-test 1`] = `
<div class="form-control block my-4">
  <div class="flex">
    <label class="label items-start px-0 py-0 cursor-pointer">
      <input type="checkbox"
             class="placeholder:text-subtle checkbox checkbox-primary w-[--vui-checkbox-size] h-[--vui-checkbox-size] rounded-[--rounded-field] border-2 bg-white checkbox-error"
             name="demo-checkbox-invalid"
      >
      <span class="label-text text-base ml-2.5">
        Yes I agree to the
        <a href="#link"
           class="link"
        >
          terms and conditions
        </a>
      </span>
    </label>
  </div>
  <label class="leading-5 mt-2 [&amp;>label]:mt-0 text-error fill-error flex"
         id="checkbox-demo-checkbox-invalid-feedback"
  >
    <div class="-mt-0.5">
      <svg xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           fill="currentColor"
           class="inline-block  w-6 h-6 mr-1.5 mb-0.5 icon-exclamation"
           aria-hidden="true"
      >
        <path d="m22.242,5.272l-3.515-3.515c-1.133-1.133-2.64-1.757-4.242-1.757h-4.971c-1.603,0-3.109.624-4.242,1.757l-3.515,3.514c-1.134,1.133-1.758,2.64-1.758,4.243v4.971c0,1.603.624,3.11,1.758,4.243l3.515,3.515c1.133,1.133,2.64,1.757,4.242,1.757h4.971c1.603,0,3.109-.624,4.242-1.757l3.515-3.514c1.134-1.133,1.758-2.64,1.758-4.243v-4.971c0-1.603-.624-3.11-1.758-4.243Zm-.242,9.213c0,1.069-.416,2.073-1.172,2.829l-3.515,3.515c-.756.755-1.76,1.171-2.828,1.171h-4.971c-1.068,0-2.072-.416-2.828-1.171l-3.515-3.515c-.756-.755-1.172-1.759-1.172-2.828v-4.971c0-1.069.416-2.073,1.172-2.829l3.515-3.515c.756-.755,1.76-1.171,2.828-1.171h4.971c1.068,0,2.072.416,2.828,1.171l3.515,3.515c.756.755,1.172,1.759,1.172,2.828v4.971Zm-11-.485V6c0-.552.447-1,1-1s1,.448,1,1v8c0,.552-.447,1-1,1s-1-.448-1-1Zm2.5,3.5c0,.828-.672,1.5-1.5,1.5s-1.5-.672-1.5-1.5.672-1.5,1.5-1.5,1.5.672,1.5,1.5Z"
              stroke="none"
        >
        </path>
      </svg>
    </div>
    You must agree
  </label>
</div>
`;

exports[`React/Checkbox LabelWithLink smoke-test 1`] = `
<div class="form-control block my-4">
  <div class="flex">
    <label class="label items-start px-0 py-0 cursor-pointer">
      <input type="checkbox"
             class="placeholder:text-subtle checkbox checkbox-primary w-[--vui-checkbox-size] h-[--vui-checkbox-size] rounded-[--rounded-field] bg-white"
             name="demo-label-link-checkbox"
      >
      <span class="label-text text-base ml-2.5">
        Yes I agree to the
        <a href="#link"
           class="link"
        >
          terms and conditions
        </a>
      </span>
    </label>
  </div>
</div>
`;
